docs/CONTRIBUTING.md

```diff
@@ -35,3 +35,47 @@ Test is determined fragile/broken when it is failing despite no defects are pres
   - Test can still be resurrected from history if needed - in ATH or plugin itself.
 
 The situation is expected to be handled in 90 days since reported. In case of no corrective action, the test will be removed from ATH.
+
+## Guidelines to avoid fragile/flaky tests
+
+### Click and wait
+
+When writing acceptance tests using Selenium it is common to click on an element and then run some other action on the resulting page.
+Most of the time the page is updated really quick (specially if there is not a network request involved), but on slow infrastructure this can take some time so that the next test action after clicking could be checking on something which is not the expected.
+
+To avoid this, use the click and wait pattern:
+
+```
+WebElement clickableElement = ...;
+clickableElement.click();
+waitFor(by.id("my-element")).withTimeout(10);
```

Comment:
you probably want to wait for something rather then the element (which may be there but invisible)?

Reply:
Implicit timeout in `find` is one second and not using `ElasticTime`. Even using ElasticTime that is good when you are running always in the same infra (ie ci.jenkins.io) and is not very dynamic as you can put a default that makes sense most of the times, but if you run the ath as part of other builds (like core or some internal builds we have, or even locally) that default timeout starts to be less usefull and you start to get flaky results... And is harder to understand to newcomers as the timeout is "hidden"

I tend to prefer this `click and wait` approach even if not allways needed for various reasons:

* Is much more easier to people reading the test understand that we expect some element to sometimes take time to appear, this makes debugging easier and IMO test intent easier
* Is more robust when running in resource constrined environments as you can put a proper timeout for the concrete operation without affecting the global speed of all tests being running by overriding the implicit global timeout

So, this is a pattern I tend to use when I need to interact with an element that I know is not already present on the page I am already in. This does not mean I recommend to put a `waitFor` allways. If the element is suposed to already be on the page the default timeout is more than enough to deal with general loading times (although it probably should be switched to use `ElasticTime` )

Reply:
> Is this needed even when they are used?

Yeah, when using `find` the waitFor is there, but I'd like to explicitly point out that the wait is generally needed. If it's implicit in your current call, good, but test devs should know (and check) that those wait calls are needed.

Anyways, I'll make a note for the find methods which already use this wait pattern.

Reply:
> which may be there but invisible

Right, fixing.

Reply:
@jtnord `CapybaraPortingLayerImpl#waitFor(By, int)` is already checking for visibility.